File: NonAbundantSums/NAS.py
#!/usr/bin/env python3

"""
Project Euler # 23


"""

import math

def factors_sum(n):
    if n == 1:
        return 1

    x = 1
    for i in range(2, int(math.sqrt(n))+1):
        if n % i == 0:
            if i*i == n:
                x += i
            else:
                x += i + (n // i)
    return x

def is_abundant(n):
    return factors_sum(n) > n

abundants = [i for i in range(1,28124) if is_abundant(i)]

sum_of_an = []
for i in abundants:
    for j in abundants:
        sum_of_an.append(i+j)
        if i + j > 28123:
            break

sum_of_an = set(sum_of_an)
print(sum(i for i in range(28124) if i not in sum_of_an))

File: NonAbundantSums/NAS.java
/*
Project Euler # 23

*/

public class NAS {
    public static void main( String[] args ) {
        int[] abundants = new int[14000];
        boolean[] sum_of_an = new boolean[28124];
        int sum = 0, lim = 28123, index = 0;

        for ( int i=2; i<lim; i++ )
            if ( is_abundant(i) )
                abundants[index++] = i;

        for ( int i=0; abundants[i] != 0; i++ )
            for ( int j=0; abundants[j] != 0; j++ ) {
                if ( abundants[i]+abundants[j] <= lim )
                    sum_of_an[abundants[i]+abundants[j]] = true;
            }

        for ( int i=0; i<=lim; i++ )
            if ( !sum_of_an[i] )
                sum += i;

        System.out.println(sum);
    }

    public static int factors_sum( int n ) {
        if ( n == 1 )
            return 1;

        int sum = 1;
        for ( int i=2; i<Math.round(Math.sqrt(n))+1; i++ )
            if ( n % i == 0 )
            {
                if ( i*i == n )
                    sum += i;
                else
                    sum += i + (n/i);
            }

        return sum;
    }

    public static boolean is_abundant( int a ) {
        return factors_sum(a) > a;
    }

}
